framework/resource/resource.go

```diff
@@ -0,0 +1,443 @@
+		// by our function plus our newly allocated state.
+		args := in.Args()
+
+		// Ensure our output marker type is set
+		if v := out.TypedSubtype(markerVal.Type, markerVal.Subtype); v != nil {
+			v.Value = markerVal.Value
+		}
+
+		// Call our function. We throw away any result types except for the error.
+		result := original.Call(args...)
+		err := result.Err()
+
+		// If the destroy was successful, we clear our state.
+		if err == nil {
+			r.initState(false)
+		}
+
+		return err
+	}, buildArgs...)
+}
+
+// initState sets the r.stateValue to a new, empty state value.
+// If zero is true, this will get set to a non-nil value. If zero is
+// false, the state will be a nil pointer type to the state type.
+func (r *Resource) initState(zero bool) {
+	if r.stateType != nil {
+		if zero {
+			r.stateValue = reflect.New(r.stateType.Elem()).Interface()
+		} else {
+			r.stateValue = reflect.New(r.stateType).Elem().Interface()
+		}
+	}
+}
+
+// loadState is the inverse of proto. This repopulates the state from the
+// serialized proto format. This will discard any previous state that is
+// currently loaded.
+func (r *Resource) loadState(s *pb.Framework_ResourceState) error {
+	// If we have no raw value in the state then ignore it.
+	if s == nil || s.Raw == nil {
+		return nil
+	}
+
+	// We try to unmarshal directly into a state value
+	r.initState(true)
+	if r.stateValue == nil {
+		return fmt.Errorf(
+			"resource %q: can't unserialize state because the resource "+
+				"has no defined state type", r.name)
+	}
+
+	pm, ok := r.stateValue.(proto.Message)
+	if !ok {
+		return fmt.Errorf(
+			"resource %q: can't unserialize state because the resource "+
+				"state type is not a protobuf message.", r.name)
+	}
+	return component.ProtoAnyUnmarshal(s.Raw, pm)
+}
+
+// proto returns the protobuf message for the state of this resource.
+func (r *Resource) proto() *pb.Framework_ResourceState {
+	stateProto, err := component.Proto(r.stateValue)
+	if err != nil {
+		// This shouldn't happen.
+		panic(err)
+	}
+
+	// This means we have no state value, we return just the name.
+	if stateProto == nil {
+		return &pb.Framework_ResourceState{Name: r.name}
+	}
+
+	// Encode our state
+	anyVal, err := component.ProtoAny(stateProto)
+	if err != nil {
+		// This shouldn't happen.
+		panic(err)
+	}
+
+	var m jsonpb.Marshaler
+	m.Indent = "\t" // make it human-readable
+	jsonVal, err := m.MarshalToString(stateProto)
+	if err != nil {
+		jsonVal = fmt.Sprintf(`{"error": %q}`, err)
+	}
+
+	return &pb.Framework_ResourceState{
+		Name: r.name,
+		Raw:  anyVal,
+		Json: jsonVal,
+	}
+}
+
+// ResourceOption is used to configure NewResource.
+type ResourceOption func(*Resource)
+
+// WithName sets the resource name. This name is used in output meant to be
+// consumed by a user so it should be descriptive but short, such as
+// "security group".
+func WithName(n string) ResourceOption {
+	return func(r *Resource) { r.name = n }
+}
```

Comment:
See previous comment, this probably shouldn't be an option because it appears we require name to be set.